### DISCUSSION TITLE
Custom action and actioncheck extension

### DISCUSSION THREAD

**Discussion by mathieu-pousse** [fail2ban/fail2ban#3452] [Q&A] [answered]
Hi there, 

I'd like to extend fail2ban and add an additional `actioncheck` before actually banning the ip.

Basically, the idea is to check in another system if the ip address is "bannable" or not (this cannot be done by using an exclusion list).

Anyway, my concern is how could I create a new action that extends the default iptables-multiport and extends (not override) the default actioncheck.

I need something like

`actioncheck = check-other-system && iptables-multiport.actioncheck`

is there a way to achieve this properly ?

Thanks in advance

**Comment by sebres** [comment, marked as answer]
> I'd like to extend fail2ban and add an additional actioncheck before actually banning the ip.

1. Don't use `actioncheck` for that. Since release v.1.0.1, 7f1b578af43faa3d54bc72aaee0c2b49c9570e25 (implementation for #488) the behavior has been changed - `actioncheck` will be invoked only in case of some error (to verify action in order to restore sane environment).
2. But it is not really necessary, because adjust of `actionban` would be fully enough. 

You can rewrite it somehow like that:
```ini
actionban = check-other-system && {%(known/actionban)s;}
```
or using `if check-other-system; then %(known/actionban)s; fi`:
```ini
actionban = if check-other-system; then
            %(known/actionban)s
            fi
```

One can even rewrite it directly in jail (without to create new action with include):
```ini
[jail]
banaction = iptables-ipset[actionban="check-other-system && {<known/actionban>;}"]
```

However if you'd also want to extend `actioncheck` (to check whether everything is sane), you can do that in the same way.
Just note that if it'd return non zero code, fail2ban would invoke `actionrepair` or if `actionrepair` is not set, would restart the action and reban all IPs hereafter.

>  the idea is to check in another system if the ip address is "bannable" or not (this cannot be done by using an exclusion list).

Why shouldn't it be "bannable"?
Anyway consider that such approach may be affected by a race condition: the IP is "bannable" to the time point of call `check-other-system`, but not "bannable" immediately hereafter so if it gets banned it could cause an error or unexpected state.
It is hardly possible to exclude such race conditions without special precautions like some (file) lock or a cross process mutex holding the lock during whole ban phase (as well as during certain mechanisms of other-system, if it can even switch "bannable" state outside of fail2ban).
Another variant to avoid race condition would be to do it vice versa: firstly try to ban and ignore in error case or if it already exists, for instance:
```ini
actionban = <iptables> -I f2b-<name> 1 -s <ip> -j <blocktype> || echo "ignore, <ip> not bannable"
```
or
```ini
actionban = ipset -exist add ... <ip>
```

**Reply by mathieu-pousse** [reply]
Thanks for the explanation 👌

Why shouldn't it be "bannable"?
I try to achieve something like: 
1/ fail2ban parses the log to detect **supsicious** ip
2/ the ban action check the "fraud score" of the ip in another database with custom rules, ....
3/ if the ip was safe until here or even was used by a "know" customer,  then we suspend the ban, otherwise the ip is actually banned (using the default iptable ban action)

for instance, when an ip is about to be banned we could do the reverse lookup to get the country and ignore abusive ips from a certain country.

I'm not sure to get your remark about the race condition.
Do you mean that "simultaneous" calls can be made for the banaction and that if, in my case, the check-other-system flips from "bannable" to "unbannable", you cannot guarantee the result wihtin fail2ban server ?

**Reply by sebres** [reply]
> the ban action check the "fraud score" of the ip in another database with custom rules, ....
> if the ip was safe until here or even was used by a "know" customer, then we suspend the ban

OK, then you are not quite correct with the approach - the ticket is anyway "banned" in fail2ban, but:
- it is not the case in the reality;
- therefore confuses fail2ban (it consider the IP as banned), e. g. also affected by ban prolongation if `bantime.increment` is enabled, etc;
- it'd be not optimal (unnecessary wasting resources).

**Use `ignorecommand` instead.** See https://github.com/fail2ban/fail2ban/issues/2013#issuecomment-426940189 for an example.

**Reply by mathieu-pousse** [reply]
:heart: looks perfect !

Thanks !
